### DISCUSSION TITLE
Clarify on policy/licensing questions for (new) stream operators

### DISCUSSION THREAD

**Discussion by kstokoz** [apache/pekko#553] [General]
Hi,

I have a question regarding adding/mirroring operators that are available in Akka streams in Pekko, i.e.

```
Source/Flow.mapAsyncPartitioned
Source/Flow.mapWithResource
Flow.contramap
```

It is allowed to add such to pekko? Are there any policies around mirroring functionality?

Regards,
Kyrylo

**Comment by pjfanning** [comment]
Apache Pekko is a fork of Akka 2.6.20. Every API that is in Akka 2.6.20 is in Pekko. If these have been newly added to Akka 2.7 or 2.8, then we can look at adding them to Pekko but we can't copy the new Akka code. Any Akka code added since 2.7.0 is licensed under a new license that is not Apache Software License compatible.

**Comment by kstokoz** [comment]
These operators were added after the fork, mapAsynPartitioned was added in 2.8 (redacted link).
How in this case it can be added to Pekko?

**Reply by pjfanning** [reply]
Noone who has looked at the Akka code can participate in writing the Pekko equivalent code. Any Pekko version of these functions would need to be clean room implementations.

**Reply by pjfanning** [reply]
I've deleted the Akka link. It is unhelpful and could get us in trouble.

**Reply by pjfanning** [reply]
I would suggest that you raise a Pekko issue but please don't submit a PR since it appears you have read the Akka code.

**Reply by pjfanning** [reply]
It may be possible that we can find out how non-Akka reactive streams implementations have created their versions of these functions (if they have such functionality). Any open source implementations of these functions could be a useful guide for how Pekko can deliver them.

**Reply by kstokoz** [reply]
Ok, thanks for the clarification.

**Reply by mdedetrich** [reply]
Note that we can look at general documentation of the function, but we are absolutely not allowed to look at the code/implementation. It has to be completely clean room implementation.

Due to this a lot of the core contributors to Pekko (including myself) have stopped looking at Akka code since the fork happened so we don't get into legal issues.

**Reply by gmethvin** [reply]
Note that this also applies to any code that is under a non-permissive license (e.g. GPL), not just Akka code. While looking at the code doesn't mean you're automatically going to violate their copyright, it increases the likelihood you will use their original ideas without realizing it, so our policy is just that you should not look at such code at all.

**Reply by gmethvin** [reply]
Note that we can incorporate code that is licensed as Apache or an Apache-compatible license such as MIT, as long as we provide proper attribution. For example you could look at this project: https://github.com/jaceksokol/akka-stream-map-async-partition

**Reply by He-Pin** [reply]
> Note that we can incorporate code that is licensed as Apache or an Apache-compatible license such as MIT, as long as we provide proper attribution. For example you could look at this project: https://github.com/jaceksokol/akka-stream-map-async-partition

May be better to ask @jaceksokol If he like to contribute his 👋

**Reply by jaceksokol** [reply]
Hi, I can create a PR with the operator, but I can't promise I will do this this week. But you're free to re-use the code.

**Reply by He-Pin** [reply]
@jaceksokol refs: https://github.com/apache/incubator-pekko/pull/561 , welcome:), on your ride.

**Reply by pjfanning** [reply]
Thanks @jaceksokol - I have already created #561. If you could review it that would be great. Or if you prefer to create a PR of your own, we can go that route too.

**Reply by jaceksokol** [reply]
The PR looks fine, the only thing is to review the licence files, but I'm not an expert to do this 😉

**Reply by He-Pin** [reply]
Maybe you can resend the PR and let's @pjfanning as a coauther, then the license is not need.

**Comment by He-Pin** [comment]
@kstokoz I can port my two operators, but still a little busy at $work,  But the `mapAsyncPartitioned` I don't have the copyright, so I can't, I don't know whether we can just implemented the API.

**Reply by pjfanning** [reply]
@He-Pin you own the copyright to the PRs you created so you can create PRs for Pekko directly from them - noone else can - because of the Akka license.

With mapAsyncPartitioned, we need to write a brand new implementation from scratch without reading the Akka code.

**Reply by He-Pin** [reply]
@pjfanning I am on my way, but My wife will come to HangZhou tomorrow, so may be a little slower , and I am on my way to clean up my pending merge queue.

And can we read the `mapAsyncPartitioned` document and examples on the website?

**Reply by pjfanning** [reply]
Please, no rush. This is useful but not worth sweating on. Any time that you have free time but do it when it suits you.

**Reply by pjfanning** [reply]
You can read the Akka issue for mapAsyncPartitioned but it best not to look at the PR. Any docs are ok too. I'm not sure about reading discussion on the PR but we shouldn't read the code - it just seems safest to stay away from the PR.

**Reply by mdedetrich** [reply]
Document/examples yes, code absolutely not.

If you implement a mirror of `mapAsyncPartitioned`, your also allowed to compare the results of your implementation versus Akka's (i.e. pass inputs into your `mapAsyncPartitioned`, take notes of the outputs and compare that to the inputs passed into Akka's `mapAsyncPartitioned`).

But again do not look at the source code of Akka when doing this, just "run" the Akka binaries (i.e. jar's).

**Reply by He-Pin** [reply]
the mapAsyncPatition is difficulty to get right, I will put it aside for now，and thanks for the clarification.

**Reply by mdedetrich** [reply]
It would be nice to have these functions at some point if we can implement it but its definitely not urgent.

**Comment by He-Pin** [comment]
@kstokoz `contramap` has been added, if you want anyother operators, please comment in https://github.com/apache/incubator-pekko/discussions/605 and let's get adequate votes.

**Comment by He-Pin** [comment]
@kstokoz https://github.com/apache/incubator-pekko/pull/931 add mapWithResource operator.
